### PR TITLE
fix: Fix BigtableIO to share one BigtableService per worker and increase default attempt timeout

### LINKED ISSUE #26166: [Bug]: Regression in BigtableIO in the new 2.47.0 versions
### What happened?

We migrated BigtableIO in https://github.com/apache/beam/pull/24015 and seems to caused regression. Dataflow jobs started failing. We're still trying to root cause it. 

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by mutianf**
fix #26166 

In https://github.com/apache/beam/pull/24015, the map that stores BigtableService is transient. During serialized and deserialized, it recreates an empty map which ended up creating a new BigtableService per each thread. The fix makes the map static. 

In the previous version, mutateRows operation has a timeout of 6 minutes. And after the migration, the default timeout became 1 minute, which makes the job fail more easily. Update the timeout back to 6 minutes. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTranslator.java [hunk 1]
```diff
@@ -180,10 +180,18 @@ private static BigtableDataSettings configureWriteSettings(
...
     RetrySettings.Builder retrySettings = callSettings.getRetrySettings().toBuilder();
     BatchingSettings.Builder batchingSettings = callSettings.getBatchingSettings().toBuilder();
+
+    // The default attempt timeout for version <= 2.46.0 is 6 minutes. Reset the timeout to align
+    // with the old behavior.
+    retrySettings
+        .setInitialRpcTimeout(Duration.ofMinutes(6))
+        .setMaxRpcTimeout(Duration.ofMinutes(6));
+
     if (writeOptions.getAttemptTimeout() != null) {
```
Comment (igorbernstein2) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableConfigTranslator.java:190]: Might be nice to clean this up a bit:

```
Duration attemptTimeout = Duration.ofMinutes(6);

if (writeOptions.getAttemptTimeout() != null) {
  attemptTimeout = Duration.ofMillis(writeOptions.getAttemptTimeout().getMillis());
}
retrySettings.setInitialRpcTimeout(attemptTimeout).setMaxRpcTime(attemptTimeout);

```

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceFactory.java [hunk 2]
```diff
@@ -44,111 +43,120 @@
...
-          factory, configId, service, refCount);
+    static BigtableServiceEntry create(ConfigId configId, BigtableService service) {
+      return new AutoValue_BigtableServiceFactory_BigtableServiceEntry(configId, service);
     }
 
     @Override
     public void close() {
-      getServiceFactory().releaseService(this);
+      int refCount =
+          refCounts.getOrDefault(getConfigId().id(), new AtomicInteger(0)).decrementAndGet();
```
Comment (igorbernstein2) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceFactory.java:75]: if refCounts doesnt contain the configId or the ref count is already < 0, please log an error

**Review by igorbernstein2 — APPROVED** [review]

**Review by jrmccluskey — APPROVED** [review]

**Comment by jrmccluskey** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by Abacn** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by jrmccluskey** [comment]
It looks like there's a somewhat consistent timeout in `testTriggeredFileLoadsWithTempTablesToExistingNullSchemaTable`:

```
17:37:22 org.apache.beam.sdk.io.gcp.bigquery.BigQueryIOWriteTest > testTriggeredFileLoadsWithTempTablesToExistingNullSchemaTable[1] FAILED
17:37:22     org.junit.runners.model.TestTimedOutException at BigQueryIOWriteTest.java:729
17:45:55 
17:45:55 1645 tests completed, 1 failed, 124 skipped
```

Is this a matter of needing higher bounds on the direct runner?

**Comment by Abacn** [comment]
@jrmccluskey This is known issue #25207. It was causing the whole GCP_IO test suite time out before I added test timeout of 10min per test. From the log the suspected cause is some racing condition that temp file gets deleted when the test run and then test hanging. It happens occassionally but recently becomes more frequent. Possibly because other tests get added to GCP_IO then the order that junit running these test changed.

**Comment by jrmccluskey** [comment]
retest this please

**Comment by jrmccluskey** [comment]
Run Java_Examples_Dataflow_Java17 PreCommit

**Comment by jrmccluskey** [comment]
Hmm still getting that Spanner IO failure. I don't have reason to believe it's related to this PR  though. CC: @johnjcasey

**Comment by Abacn** [comment]
> Hmm still getting that Spanner IO failure. I don't have reason to believe it's related to this PR though. CC: @johnjcasey

That is #26208 unrelated to Beam

**Pull request merged**